packages/tabs/src/Tabs.tsx

```diff
@@ -163,105 +173,116 @@ function Tabs({
     ? setControlledSelected
     : setUncontrolledSelected;
 
-  function handleChange(
-    e: React.SyntheticEvent<Element, MouseEvent>,
-    index: number,
-  ) {
-    setSelected(index);
-  }
+  const handleChange = useCallback(
+    (e: React.SyntheticEvent<Element, MouseEvent>, index: number) => {
+      setSelected(index);
+    },
+    [setSelected],
+  );
 
-  const getEnabledIndexes: () => [Array<number>, number] = () => {
+  const getEnabledIndexes: () => [Array<number>, number] = useCallback(() => {
     const enabledIndexes = childrenArray
       .filter(child => !child.props.disabled)
       .map(child => childrenArray.indexOf(child));
 
     return [enabledIndexes, enabledIndexes.indexOf(selected!)];
-  };
-
-  const handleArrowKeyPress = (e: KeyboardEvent) => {
-    if (!(e.metaKey || e.ctrlKey)) {
-      if (e.keyCode === keyMap.ArrowRight) {
-        const [enabledIndexes, current] = getEnabledIndexes();
-        setSelected(enabledIndexes[(current + 1) % enabledIndexes.length]);
-      } else if (e.keyCode === keyMap.ArrowLeft) {
-        const [enabledIndexes, current] = getEnabledIndexes();
-        setSelected(
-          enabledIndexes[
-            (current - 1 + enabledIndexes.length) % enabledIndexes.length
-          ],
-        );
+  }, [childrenArray, selected]);
+
+  const handleArrowKeyPress = useCallback(
+    (e: KeyboardEvent) => {
+      if (!(e.metaKey || e.ctrlKey)) {
+        if (e.keyCode === keyMap.ArrowRight) {
+          const [enabledIndexes, current] = getEnabledIndexes();
+          setSelected(enabledIndexes[(current + 1) % enabledIndexes.length]);
+        } else if (e.keyCode === keyMap.ArrowLeft) {
+          const [enabledIndexes, current] = getEnabledIndexes();
+          setSelected(
+            enabledIndexes[
+              (current - 1 + enabledIndexes.length) % enabledIndexes.length
+            ],
+          );
+        }
       }
-    }
-  };
+    },
+    [getEnabledIndexes, setSelected],
+  );
 
-  const tabs = React.Children.map(children, (child, index) => {
-    if (!isComponentType<'Tab'>(child, 'Tab')) {
-      return child;
-    }
+  const { tabs, tabpanels } = useMemo(() => {
+    const tabs: Array<React.ReactElement> = [];
 
-    return React.cloneElement(child, {
-      key: index,
-      ariaControl: `tab-${index}`,
-      selected: selected === index,
+    const tabpanels = React.Children.map(children, (child, index) => {
+      if (!isComponentType<'Tab'>(child, 'Tab')) {
+        return child;
+      }
+
+      const { id: idProp, name, disabled, onClick } = child.props;
+      const tabpanelId = idProp ?? tabpanelIdAllocator.generate();
+      const tabId = idProp ?? tabIdAllocator.generate();
```

Comment:
when `idProp` is defined, both the `Tab` and `TabTitle` will have the same ID, but IDs should be unique, so it seems like `tabpanelId` should always be generated. if `idProp` is not defined, these will generate new ids on every render. why don't `Tab` and `TabTitle` generate their own IDs that they can memoize? can we get the values for `aria-controls` and `aria-labelledby` from `Ref`s?

Reply:
Good catch on the tabpanelId, but I'm not sure that it necessarily makes more sense to have Tab and TabTitle generate their own ids and then relate them to one another through refs. If we're already mapping through the Tabs' children and creating the TabTitles based on those Tab components, we may as well relate them while we do so.

Reply:
alternatively, can `TabTitle` handle generating all the IDs and setting the props on its child `Tab`? or can we create an `InternalTab` component which creates the `TabTitle` and generates the IDs

Reply:
TabTitle doesn't know about its child Tab, it's strictly responsible for rendering the button that appears in the tablist, relating the two anywhere but inside of Tabs gets complicated because it is in Tabs that we create the TabTitle from the Tab.

Reply:
If I make an InternalTabs or InternalTab component, would it be responsible for doing anything outside of the logic contained from lines 213-259?

Reply:
I would say 218-259. by being it's only component it would be able to use hooks like memo

Reply:
~30 lines of that are defining the TabTitle component 😱 so it seems like the main benefit we'd be getting here is just in memoizing the ids independently of the large memo that they are currently a part of?

Reply:
I'm happy to make the change, I just want to make sure the benefits are worth the tradeoff of breaking up the logic and making the file a little less readable

Reply:
theoretically we could clean up the large memo by wrapping `InternalTab` in `React.memo`